[docs] Document NativeAOT support. Fixes #18585. (#19362)
xamarin/xamarin-macios: docs/nativeaot.md

# NativeAOT

We've added *experimental* support for using [NativeAOT][1] when publishing iOS,
tvOS, Mac Catalyst and macOS apps in .NET 8.

NativeAOT may produce smaller and/or faster apps - or it may not. It's very
important test and profile to determine the results of enabling NativeAOT.

> 

Reviewer: ```suggestion

## How to enable NativeAOT?

```

> NativeAOT is enabled by setting the `PublishAot` property to `true` in the project file:

Reviewer 2: It's not necessary to set `$(PublishAotUsingRuntimepack)=true` anymore?

Reviewer: Afaik it is still required if you do not want to change your project file but just force defaults through CLI, e.g.,:
```
dotnet publish -c Release -f net8.0-ios -r ios-arm64 -p:PublishAot=true -p:PublishAotUsingRuntimePack=true
```

However, passing `-p:PublishAot=true` on the command-line is discouraged in general, as it will not just enable NativeAOT-targets, but has some additional dependencies in the SDK to run analyzers, disable generating runtimeconfig.json,...  So I suppose it is OK that we do not mention it and avoid confusion.

Regarding the project-level setting, Xamarin SDK helps by setting the property for the user:
https://github.com/xamarin/xamarin-macios/blob/2534418caa473fc7e869095e220329230866ff41/dotnet/targets/Xamarin.Shared.Sdk.props#L179-L185

Author: I'm not sure I understand why it would be different when passing PublishAot from the command line vs in the project?

Reviewer: > I'm not sure I understand why it would be different when passing PublishAot from the command line vs in the project?

There are issues with the framework resolution in the SDK, when we have multiple target frameworks specified in the project file (e.g., MAUI projects). 
Building a MAUI app without adding `PublishAot` to the project file, but passing it through command line:

```
dotnet publish -f net8.0-ios -c Release -r ios-arm64 -p:PublishAot=true
```

Gives the following error - coming from the SDK:
```
/usr/local/share/dotnet/sdk/8.0.100-rc.2.23502.2/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.Sdk.FrameworkReferenceResolution.targets(90,5): error NETSDK1203: Ahead-of-time compilation is not supported for the target runtime identifier 'ios-arm64'. [/Users/ivan/tmp/AppTest/AppTest.csproj::TargetFramework=net8.0-android]
```

Including the `-p:PublishAotUsingRuntimePack=true` in the command line above, works around the issue.

--- 
Additionally, regarding what I said about passing `PublishAot` only through command line being "discouraged", is the fact that `dotnet build` and `dotnet publish` with and without `PublishAot=true` mean different things, which can be confusing to the end user. 
If we consider a desktop app doing the following:

```csharp
var asm = System.Reflection.Assembly.LoadFrom("somepath");
``` 
1. When built with `dotnet build` - just builds the app as a regular managed desktop app:
    - outputs:
    ```
    MSBuild version 17.8.0+6cdef4241 for .NET
    Determining projects to restore...
    Restored /Users/ivan/tmp/NaotApp/NaotApp.csproj (in 78 ms).
    /usr/local/share/dotnet/sdk/8.0.100-rc.2.23502.2/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.RuntimeIdentifierInference.targets(311,5): message NETSDK1057: You are using a preview version of .NET. See: https://aka.ms/dotnet-support-policy [/Users/ivan/tmp/NaotApp/NaotApp.csproj]
    NaotApp -> /Users/ivan/tmp/NaotApp/bin/Debug/net8.0/NaotApp.dll

    Build succeeded.
    0 Warning(s)
    0 Error(s)
    ```

2. When built with `dotnet build -p:PublishAot=true` - builds the app as a managed desktop app with trim/AOT analyzers enabled (indicating it might be published/deployed afterwards):
    - outputs:
    ```
    MSBuild version 17.8.0+6cdef4241 for .NET
    Determining projects to restore...
    Restored /Users/ivan/tmp/NaotApp/NaotApp.csproj (in 127 ms).
    /usr/local/share/dotnet/sdk/8.0.100-rc.2.23502.2/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.RuntimeIdentifierInference.targets(311,5): message NETSDK1057: You are using a preview version of .NET. See: https://aka.ms/dotnet-support-policy [/Users/ivan/tmp/NaotApp/NaotApp.csproj]
    /Users/ivan/tmp/NaotApp/Program.cs(1,11): warning IL2026: Using member 'System.Reflection.Assembly.LoadFrom(String)' which has 'RequiresUnreferencedCodeAttribute' can break functionality when trimming application code. Types and members the loaded assembly depends on might be removed. [/Users/ivan/tmp/NaotApp/NaotApp.csproj]
    NaotApp -> /Users/ivan/tmp/NaotApp/bin/Debug/net8.0/NaotApp.dll

    Build succeeded.

    /Users/ivan/tmp/NaotApp/Program.cs(1,11): warning IL2026: Using member 'System.Reflection.Assembly.LoadFrom(String)' which has 'RequiresUnreferencedCodeAttribute' can break functionality when trimming application code. Types and members the loaded assembly depends on might be removed. [/Users/ivan/tmp/NaotApp/NaotApp.csproj]
    1 Warning(s)
    0 Error(s)
    ```
3. When built with `dotnet publish` - deploys the app without enabling trim/AOT analyzers (although I am not 100% sure in which mode, I would suppose it is still a framework dependent one)
    - output:
    ```
    MSBuild version 17.8.0+6cdef4241 for .NET
    Determining projects to restore...
    Restored /Users/ivan/tmp/NaotApp/NaotApp.csproj (in 78 ms).
    /usr/local/share/dotnet/sdk/8.0.100-rc.2.23502.2/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.RuntimeIdentifierInference.targets(311,5): message NETSDK1057: You are using a preview version of .NET. See: https://aka.ms/dotnet-support-policy     [/Users/ivan/tmp/NaotApp/NaotApp.csproj]
    NaotApp -> /Users/ivan/tmp/NaotApp/bin/Release/net8.0/NaotApp.dll
    NaotApp -> /Users/ivan/tmp/NaotApp/bin/Release/net8.0/publish/
    ```
4. When built with `dotnet publish -p:PublishAot=true` - deploys the app with enabled trim/AOT analyzers and with NativeAOT
    - output:
    ```
    MSBuild version 17.8.0+6cdef4241 for .NET
    Determining projects to restore...
    Restored /Users/ivan/tmp/NaotApp/NaotApp.csproj (in 149 ms).
    /usr/local/share/dotnet/sdk/8.0.100-rc.2.23502.2/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.RuntimeIdentifierInference.targets(311,5): message     NETSDK1057: You are using a preview version of .NET. See: https://aka.ms/dotnet-support-policy     [/Users/ivan/tmp/NaotApp/NaotApp.csproj]
    /Users/ivan/tmp/NaotApp/Program.cs(1,11): warning IL2026: Using member 'System.Reflection.Assembly.LoadFrom(String)' which has 'RequiresUnreferencedCodeAttribute' can break functionality when trimming application code. Types and members the loaded assembly depends on might be removed. [/Users/ivan/tmp/NaotApp/NaotApp.csproj]
    NaotApp -> /Users/ivan/tmp/NaotApp/bin/Release/net8.0/osx-arm64/NaotApp.dll
    Generating native code
    /Users/ivan/tmp/NaotApp/Program.cs(1): Trim analysis warning IL2026: Program.<Main>$(String[]): Using member 'System.Reflection.Assembly.LoadFrom(String)' which has 'RequiresUnreferencedCodeAttribute' can break functionality when trimming application code. Types and members the loaded assembly depends on might be removed. [/Users/ivan/tmp/NaotApp/NaotApp.csproj]
    NaotApp -> /Users/ivan/tmp/NaotApp/bin/Release/net8.0/osx-arm64/publish/
    ```
5. Additionally, when built with `dotnet publish -p:PublishReadyToRun=true` - deploys the app without enabling trim/AOT analyzers and with ReadyToRunCrossgen2
    - output:
    ```
    MSBuild version 17.8.0+6cdef4241 for .NET
    Determining projects to restore...
    Restored /Users/ivan/tmp/NaotApp/NaotApp.csproj (in 9.55 sec).
    /usr/local/share/dotnet/sdk/8.0.100-rc.2.23502.2/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.RuntimeIdentifierInference.targets(311,5): message     NETSDK1057: You are using a preview version of .NET. See: https://aka.ms/dotnet-support-policy     [/Users/ivan/tmp/NaotApp/NaotApp.csproj]
    NaotApp -> /Users/ivan/tmp/NaotApp/bin/Release/net8.0/osx-arm64/NaotApp.dll
    NaotApp -> /Users/ivan/tmp/NaotApp/bin/Release/net8.0/osx-arm64/publish/
    ```
    
Because of all these differences when using `dotnet build` or `dotnet publish` (and various deployment modes), I believe the principle is: `If you want to use NativeAOT - set the opt-in property in the project file`
This gives you: 
- necessary information (trim/AOT warnings) for the inner-dev loop experience - `dotnet build` to guide how you write your code so it is trim/NativeAOT compatible 
- on `dotnet publish` the deployment mode compatible with step 1. 
- finally, a simple way of using NativeAOT ie without needing to understand all other modes and concepts 

I know I gave an example for desktop platforms, but for mobile we also have similar confusion for Xamarin legacy / Xamarin .NET, MonoAOT / MonoAOT+interpreter, MonoAOT vs NativeAOT, so I suppose for simplicity we should try to follow the same principle, at least until we do not solve SDK issues like the need for `PublishAotUsingRuntimePack` and https://github.com/dotnet/sdk/issues/25392 

---

Hope this helps :)

Reviewer 2: @ivanpovazan I think that we should focus just on the csproj settings here in this document. Does it mean that since we suggest adding `<PublishAot>true</PublishAot>` to the .csproj, we should also guide the devs to put `<PublishAotUsingRuntimepack>true</PublishAotUsingRuntimepack>` next to it?

Reviewer: If `PublishAot` is specified in the project file, `PublishAotUsingRuntimepack` is not needed.

```xml
<PropertyGroup>
	<PublishAot>true</PublishAot>
</PropertyGroup>
```

## Notes

NativeAOT is only used when publishing (`dotnet publish`). In particular
`dotnet build -t:Publish` is _not_ equivalent to `dotnet publish`, only
`dotnet publish` will enable NativeAOT.

**Unsupported** workaround: set the `_IsPublishing` property to `true` to make
`dotnet build` think it's `dotnet publish`:

```xml
<PropertyGroup>
	<PublishAot>true</PublishAot>
	<_IsPublishing>true</_IsPublishing>
</PropertyGroup>
```

This can be useful to install and run apps with `NativeAOT` from the IDE,
because it's not possible to use `dotnet publish` when running from the IDE.

> 

Reviewer: ```suggestion

## Compatibility and limitations

```

> Note that NativeAOT does not support managed debugging.
> 
> There's no interpreter when using NativeAOT, and as such the `UseInterpreter`
> and `MtouchInterpreter` properties have no effect.
> 
> NativeAOT requires trimming, and `MAUI` isn't trimmer-safe, and thus
> unfortunately `MAUI` projects don't typically work with NativeAOT (we hope to
> rectify this situation for .NET 9).
> 
> There are no known issues specific to our platforms with NativeAOT; but the
> [limitations][2] are exactly the same as for other platforms.

Reviewer: ```suggestion
There are no known issues specific to our platforms with NativeAOT; but the
[limitations][2] are exactly the same as for other supported platforms. 
Nevertheless, we would like to point out a few features that are not available with NativeAOT, that are with Mono, when targeting Apple platforms:

- NativeAOT does not support managed debugging.
- There's no interpreter when using NativeAOT, and as such the `UseInterpreter`
and `MtouchInterpreter` properties have no effect.
- NativeAOT requires trimming, and `MAUI` isn't trimmer-safe, and thus
unfortunately `MAUI` projects don't typically work with NativeAOT (we hope to
rectify this situation for .NET 9).
```

Reviewer: @rolfbjarne regarding the issue with MAUI not being trimmer-safe, do we want to mention that setting `TrimMode=partial` could help?

On that note, should we maybe generalise the problem to all non-trimmable assemblies?

Author: I never tested MAUI projects, but I was under the impression they didn't work, no workaround possible?

Reviewer: I will take a look, I believe I had an example where `TrimMode=partial` helped, and still produced smaller app than Mono - I will double-check.
I just felt that `don't typically work` is a bit harsh (even though its true).

Reviewer: Here is the gist with analysis of building/running 5 different MAUI iOS apps with NativeAOT using different trimming modes: https://gist.github.com/ivanpovazan/4756bdb65a34366f0944e99f601379f3 

I did not go deeper with examining why something failed, but just wanted to give an overview regarding `TrimMode=partial` vs `TrimMode=full`

# NativeAOT

We've added *experimental* support for using [NativeAOT][1] when publishing iOS,
tvOS, Mac Catalyst and macOS apps in .NET 8.

NativeAOT may produce smaller and/or faster apps - or it may not. It's very
important to test and profile to determine the results of enabling NativeAOT.

However, our initial testing shows significant improvements both in size (up
to 50% smaller) and startup (up to 50% faster). For more information about
performance see [.NET 8 Performance Improvements in .NET MAUI][3].

> 

Reviewer 3: This is not exactly confidence inspiring. If I were reading this, I would like to know under what circumstances this would generate larger or slower apps rather than leaving it up to me to profile.

Author: Unfortunately there's no easy way to predict how an app will turn out. The fact that projects are trimmed (linked) means that adding a single line of code somewhere can have a tremendous cascading effect, preventing *a lot* of other code from being removed. Then add the effect of different AOT compilers (Mono's AOT might generate bigger code in particular when it comes to generics, but there might be other code patterns that may have the opposite effect) multiplied by whatever code patterns is used by all the code that was brought in by that single line of extra code in the project... the answer really is that each project is different, and that developers must test their projects to see the effect of turning on NativeAOT. I'm not sure how to phrase it in a way that inspires more confidence?

Reviewer: @rolfbjarne maybe something like this might help you phrase it in the right way:

```suggestion
Our initial testings shows significant improvements both in size (up
to 50% smaller) and startup (up to 50% faster) compared to the default offering with the Mono runtime (for more information please checkout [.NET 8 Performance Improvements in .NET MAUI][3]). 
However, it is important to point out that NativeAOT requires trimming whole applications `<MtouchLink>Full</MtouchLink>` (enabled by default) and is much more restrictive and aggressive in removing code which is not AOT compatible. As a result, if the applications  is referencing some AOT incompatible constructs (e.g., dynamically referencing code for which a dependency cannot be statically determined), NativeAOT compiler might remove them during build time, which can result with crashes at runtime. To warn users about when this can happen, adequate trimming and AOT warnings are produced and they should not be neglected. Instead, all such warnings should be addressed, code adapted accordingly, application tested and profiled in this configuration. Only then it is safe to assume that the application is compatible with NativeAOT and should not cause any unexpected behaviour. 
Finally, there could be cases when fixing mentioned trimming and AOT warnings is not possible i.e. warnings coming from referenced frameworks or third-party libraries. In such cases, `<MtouchLink>SdkOnly</MtouchLink>` can be specified on the project level as a workaround. This does affect the application size, but even in this mode NativeAOT tends to produce up to 28% smaller applications compared to Mono.
```

Author: I updated it according to Ivan's suggestion, @stephen-hawley does it look better now?

```xml
<PropertyGroup>
	<PublishAot>true</PublishAot>
	<_IsPublishing>true</_IsPublishing>
</PropertyGroup>
```

This can be useful to install and run apps with `NativeAOT` from the IDE,
because it's not possible to use `dotnet publish` when running from the IDE.

## Compatibility and limitations

There are no known issues specific to our platforms with NativeAOT; but the
[limitations][2] are exactly the same as for other supported platforms.

Nevertheless, we would like to point out a few features that are not available
with NativeAOT, that are with Mono, when targeting Apple platforms:

- NativeAOT does not support managed debugging.

- There's no interpreter when using NativeAOT, and as such the
  `UseInterpreter` and `MtouchInterpreter` properties have no effect.

- NativeAOT requires trimming, and `MAUI` isn't trimmer-safe, and thus
  unfortunately `MAUI` projects don't typically work with NativeAOT (we hope
  to rectify this situation for .NET 9).

[1]: https://learn.microsoft.com/en-us/dotnet/core/deploying/native-aot
[2]: https://learn.microsoft.com/en-us/dotnet/core/deploying/native-aot/?tabs=net8plus%2Cwindows#limitations-of-native-aot-deployment

> [3]: https://devblogs.microsoft.com/dotnet/pending_link_to_blog_post

Reviewer: ```suggestion
[3]: https://devblogs.microsoft.com/dotnet/dotnet-8-performance-improvements-in-dotnet-maui/
```
